core/src/HIP/Kokkos_HIP_Space.cpp

```diff
@@ -195,6 +212,66 @@ void* HIPHostPinnedSpace::impl_allocate(
 
   return ptr;
 }
+
+void* HIPManagedSpace::allocate(const size_t arg_alloc_size) const {
+  return allocate("[unlabeled]", arg_alloc_size);
+}
+void* HIPManagedSpace::allocate(const char* arg_label,
+                                const size_t arg_alloc_size,
+                                const size_t arg_logical_size) const {
+  return impl_allocate(arg_label, arg_alloc_size, arg_logical_size);
+}
+void* HIPManagedSpace::impl_allocate(
+    const char* arg_label, const size_t arg_alloc_size,
+    const size_t arg_logical_size,
+    const Kokkos::Tools::SpaceHandle arg_handle) const {
+  void* ptr = nullptr;
+
+  if (arg_alloc_size > 0) {
+    if (++num_hip_managed_allocations == 0 && Kokkos::show_warnings()) {
+      if (!Kokkos::Impl::hip_driver_check_page_migration(m_device)) {
+        std::cerr
+            << R"warning(The combination of device and system configuration does not support page migration between device and host. HIPManagedSpace might not work as expected. Please refer to the Rocm documentation on unified/managed memory\n)warning"
+            << std::endl;
+      }
+
+      // check for correct runtime environment
+      const char* hsa_xnack = std::getenv("HSA_XNACK");
+      if (!hsa_xnack)
+        std::cerr
+            << R"warning(Kokkos did not find an environment variable 'HSA_XNACK' for the current process. Nevertheless, xnack is enabled for all processes if amdgpu.noretry=0 was set in the Linux kernel boot line. Without xnack enabled, Kokkos::HIPManaged might not behave as expected.)warning"
+            << std::endl;
+      else if (Kokkos::Impl::strcmp(hsa_xnack, "1"))
+        std::cerr
+            << R"warning(Kokkos detected the environement variable 'HSA_XNACK=)warning"
+            << hsa_xnack
+            << R"warning('. Kokkos advises to set it to '1' to enable it per process.)warning"
+            << std::endl;
+    }
+    auto const error_code = hipMallocManaged(&ptr, arg_alloc_size);
+    if (error_code != hipSuccess) {
+      // This is the only way to clear the last error, which we should do here
+      // since we're turning it into an exception here
+      (void)hipGetLastError();
+      throw HIPRawMemoryAllocationFailure(
+          arg_alloc_size, error_code,
+          RawMemoryAllocationFailure::AllocationMechanism::HIPMallocManaged);
+    }
+    KOKKOS_IMPL_HIP_SAFE_CALL(hipMemAdvise(
+        ptr, arg_alloc_size, hipMemAdviseSetCoarseGrain, m_device));
+
+    ++num_hip_managed_allocations;
```

Comment:
```suggestion
```

Reply:
As we do not provide an interface to the number of managed allocations, I switched to a `std::atomic<bool>` in [8c7618e](https://github.com/kokkos/kokkos/pull/5112/commits/8c7618e181b9b6b5f814e15be536f6163167d4d5)